Software/grab/DDuplGrabber.cpp

```diff
@@ -0,0 +1,480 @@
+    }
+
+    return GrabResultOk;
+}
+
+GrabResult DDuplGrabber::grabScreens()
+{
+    if (m_state != Allocated)
+    {
+        if (m_state == AccessDenied)
+        {
+            // If Access is denied, we are on a secure desktop or a 3D application is running
+            // Return black buffers and retry allocation in isReallocationNeeded
+            return returnBlackBuffer();
+        }
+        else
+        {
+            return GrabResultFrameNotReady;
+        }
+    }
+
+    try
+    {
+        bool anyUpdate = false;
+        for (GrabbedScreen& screen : _screensWithWidgets)
+        {
+            if (screen.associatedData == NULL)
+            {
+                return GrabResultError;
+            }
+
+            DDuplScreenData* screenData = (DDuplScreenData*)screen.associatedData;
+            DXGI_OUTDUPL_FRAME_INFO frameInfo;
+            IDXGIResourcePtr resource;
+            HRESULT hr = screenData->duplication->AcquireNextFrame(ACQUIRE_TIMEOUT_INTERVAL, &frameInfo, &resource);
+            if (hr == DXGI_ERROR_WAIT_TIMEOUT)
+            {
+                continue;
+            }
+            else if (hr == DXGI_ERROR_ACCESS_LOST || hr == DXGI_ERROR_INVALID_CALL)
+            {
+                // in theory, DXGI_ERROR_INVALID_CALL is returned if the frame was not released
+                // it also happens in conjunction with secure desktop (even though the frame was properly released)
+                m_state = LostAccess;
+                qWarning(Q_FUNC_INFO " Lost Access to desktop 0x%X: 0x%X, requesting realloc", screen.screenInfo.handle, hr);
+                return GrabResultFrameNotReady;
+            }
+            else if (FAILED(hr))
+            {
+                qCritical(Q_FUNC_INFO " Failed to AcquireNextFrame: 0x%X", hr);
+                return GrabResultError;
+            }
+            anyUpdate = true;
+
+            ID3D11Texture2DPtr texture;
+            hr = resource->QueryInterface(IID_ID3D11Texture2D, (void**)&texture);
+            if (FAILED(hr))
+            {
+                qCritical(Q_FUNC_INFO " Failed to cast resource to ID3D11Texture2D: 0x%X", hr);
+                return GrabResultError;
+            }
+
+            D3D11_TEXTURE2D_DESC desc;
+            texture->GetDesc(&desc);
+
+            if (desc.Width != screen.screenInfo.rect.width() || desc.Height != screen.screenInfo.rect.height())
+            {
+                qCritical(Q_FUNC_INFO " Dimension mismatch: screen %d x %d, texture %d x %d",
+                    screen.screenInfo.rect.width(),
+                    screen.screenInfo.rect.height(),
+                    desc.Width,
+                    desc.Height);
+                return GrabResultError;
+            }
+
+            size_t sizeNeeded = desc.Height * desc.Width * 4; // Assumes 4 bytes per pixel
+            if (screen.imgData == NULL)
+            {
+                screen.imgData = (unsigned char*)malloc(sizeNeeded);
```

Comment:
@psieg , how do you think, is it possible to avoid allocating our buffer for the texture pixels?
It seems, that simple copying of the texture data to our buffer can eat a lot of CPU.
For example, see PR in my own fork for MacOS, https://github.com/dreamer-dead/MyLightpack/pull/2/files#diff-bb7f091bc59772f2f10989d037e6866cL139
It lowered CPU consumption on my machine.
For systems with 4K display resolution unpacked texture data can take a lot of memory.